lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -60,6 +65,43 @@ def __init__(self, *args, distro: Type[OpenSearchDistribution] = None):
         self.secrets = SecretStore(self)
         self.tls = OpenSearchTLS(self, TLS_RELATION)
 
+    def on_allocation_exclusion_add_failed(self):
+        """Callback for when the OpenSearch service fails stopping."""
+        self.unit.status = BlockedStatus(AllocationExclusionFailed)
+
+    def on_unassigned_shards(self, unassigned_shards: int):
+        """Called during node shutdown / horizontal scale-down if some shards left unassigned."""
+        self.app.status = MaintenanceStatus(HorizontalScaleUpSuggest.format(unassigned_shards))
+
+    def append_allocation_exclusion_to_remove(self, unit_name) -> None:
+        """Store a unit in the relation data bag, to be removed from the allocation exclusion."""
+        if not self.unit.is_leader():
+            self.unit_peers_data["remove_from_allocation_exclusions"] = unit_name
+            return
+
+        exclusions = set(
+            self.app_peers_data.get("remove_from_allocation_exclusions", "").split(",")
+        )
+        exclusions.add(unit_name)
+
+        self.app_peers_data["remove_from_allocation_exclusions"] = ",".join(exclusions)
+
+    def remove_allocation_exclusions(self, exclusions: Set[str]) -> None:
+        """Remove the allocation exclusions from the peer databag if existing."""
+        stored_exclusions = set(
+            self.app_peers_data.get("remove_from_allocation_exclusions", "").split(",")
+        )
+        exclusions_to_keep = ",".join(stored_exclusions - exclusions)
+
+        if self.unit.is_leader():
+            self.app_peers_data["remove_from_allocation_exclusions"] = exclusions_to_keep
+        else:
+            self.unit_peers_data["remove_from_allocation_exclusions"] = exclusions_to_keep
```

Comment:
I don't know enough about OpenSearch, but is ok to have a `set` with multiple exclusions here? I saw that on L79 only one unit name is saved to this field in the unit peers data. Maybe a set with multiple exclusions will not reach this part of the code, so in that case ignore this comment.

Reply:
Yes, so you can set 1 single exclusion (like the one with the unit_name), but in order to have multiple exclusions, they must be set as a comma separated string, i.e: "unit1, unit2, unit3..."